### PR TITLE
OpenFDA FAERS pipeline integration

### PR DESCRIPTION
**Pull request by mbdebian**
This pull request integrates OpenFDA FAERS data processing from the independent [OpenFDA ETL pipeline](https://github.com/opentargets/platform-etl-openfda-faers) into the main ETL pipeline.

### CODE REVIEW

**Review by cmalangone — CHANGES_REQUESTED** [review]
There are already few methods for reading the input files and generating (writing) the output file.

**Comment by mbdebian** [comment]
I have refactor part of the code to properly integrate OpenFDA FAERS pipeline into the main ETL.
I still need to get up to date data for a testing run.
Could you guys have a look at the code to give me some feedback on things I may be missing?
Thanks!

File: src/main/scala/io/opentargets/etl/backend/OpenFda.scala [hunk 1]
```diff
@@ -0,0 +1,81 @@
...
+    if (context.configuration.openfda.sampling.enabled) {
+      StratifiedSampling(dfsData(FdaData()).data, fdaDataWithSummaryStats, fdaDataWithMeddra)
+    }
+    // Compute Montecarlo Sampling
+    val montecarloResults = MonteCarloSampling(
+      fdaDataWithMeddra,
+      context.configuration.openfda.montecarlo.percentile,
+      context.configuration.openfda.montecarlo.permutations
+    ).persist(StorageLevel.MEMORY_AND_DISK_SER)
+    // TODO - Produce Output
```
Comment (cmalangone) [inline src/main/scala/io/opentargets/etl/backend/OpenFda.scala:73]: Remove this comment or change it ?

**Review by JarrodBaker — CHANGES_REQUESTED** [review]
Just some quick notes before I go through it more completely: 
- The readme hasn't been updated
- There are no tests (there are some basic ones in the old repository that can be ported over): you would have noticed that we're not aiming for 100% ;)
- We don't sign code with signature blocks: we have git-blame for that.
I'm meeting with Cinzia this afternoon and I'll find out what her concerns were then. I'll go through this properly this afternoon hopefully.

**Comment by cmalangone** [comment]
> Just some quick notes before I go through it more completely:
> 
> * The readme hasn't been updated
> * There are no tests (there are some basic ones in the old repository that can be ported over): you would have noticed that we're not aiming for 100% ;)
> * We don't sign code with signature blocks: we have git-blame for that.
>   I'm meeting with Cinzia this afternoon and I'll find out what her concerns were then. I'll go through this properly this afternoon hopefully.

The loaders and writers methods are fine by me. I recall @JarrodBaker wrote some tests for this matter so it would be great if we can bring them in.

**Comment by mbdebian** [comment]
> Just some quick notes before I go through it more completely:
> 
> * The readme hasn't been updated
> * There are no tests (there are some basic ones in the old repository that can be ported over): you would have noticed that we're not aiming for 100% ;)
> * We don't sign code with signature blocks: we have git-blame for that.
>   I'm meeting with Cinzia this afternoon and I'll find out what her concerns were then. I'll go through this properly this afternoon hopefully.

I'll remove the signature blocks, they were already set in my IDE configuration from previous work on microservices in identifiers.org.

Regarding the tests, I'll have a look at the old repository and port them to this pull request.

Thanks,
Manuel

File: src/main/resources/reference.conf [hunk 2]
```diff
@@ -981,3 +982,66 @@ search {
...
+        permutations: 100
+        percentile: 0.95
+    }
+    sampling { // --- DONE ---
+        size = 0.1
+        enabled = false
+    }
+    outputs = {
+        fda-unfiltered {
+            format = "parquet"
```
Comment (JarrodBaker) [inline src/main/resources/reference.conf:1034]: The output format should be set using the variable `common.output-format`.

File: src/main/resources/reference.conf [hunk 3]
```diff
@@ -981,3 +982,66 @@ search {
...
 }
+
+// OpenFDA FAERS Configuration
+
+openfda {
+    // NOTE - Each step seems to have a commong baseline path that should be refactored
+    step-root-input-path = ${common.input}"/fda"
+    step-root-output-path = ${common.output}"/fda"
+    // Source data
+    chembl-drugs { // --- DONE ---
```
Comment (JarrodBaker) [inline src/main/resources/reference.conf:993]: Are these 'DONE' flags development artifacts?

File: src/main/scala/io/opentargets/etl/backend/openfda/stage/PrepareDrugList.scala [hunk 4]
```diff
@@ -0,0 +1,32 @@
...
+      .withColumn("drug_names",
+        array_distinct(
+          flatten(array(col("trade_names"), array(col("pref_name")), col("synonyms")))))
+      .withColumn("_drug_name", explode(col("drug_names")))
+      .withColumn("drug_name", lower(col("_drug_name")))
+      .select("chembl_id", "drug_name")
+      .distinct()
+      .orderBy(col("drug_name"))
+
+    // NOTE - I can probably remove this and leave just the operation on dfChembl?
```
Comment (JarrodBaker) [inline src/main/scala/io/opentargets/etl/backend/openfda/stage/PrepareDrugList.scala:29]: You can remove it, but it's often useful to debugging to have it structured like this as you can set a breakpoint on.

File: src/main/scala/io/opentargets/etl/backend/OpenFda.scala (hunk 1 above)
Comment (mbdebian) [reply src/main/scala/io/opentargets/etl/backend/OpenFda.scala]: Done

File: src/main/resources/reference.conf (hunk 2 above)
Comment (mbdebian) [reply src/main/resources/reference.conf]: Done

File: src/main/resources/reference.conf (hunk 3 above)
Comment (mbdebian) [reply src/main/resources/reference.conf]: Removed

**Comment by mbdebian** [comment]
Tests are now ported and working (thanks @JarrodBaker for the extra pair of eyes), and README is up to date with information related to the new ETL step.

**Review by JarrodBaker — APPROVED** [review]

**Pull request merged**